# 【自前python講座-002】組み込み関数について

---

- ref: https://docs.python.org/ja/3.7/library/functions.html
- blog: https://slash-z.com/
- github: https://github.com/KazutoMakino/PythonCourses

インポートや定義しなくても使えるような，元から python にて用意されている組み込み関数は，python 3.9 においては全部で 69 種類あります．  
この中にはあまり使わない関数もあるため，良く使われる 37 種類の組み込み関数について紹介いたします．

---

## 0. 目次: <a class="anchor" id="0"></a>

| idx | title || idx | title |
| --- | ----- |-| --- | ----- |
| 1. | [print(): 出力関数](#print) || 21. | [len(): 要素数／長さを取得](#len) |
| 2. | [input(): 入力関数](#input) || 22. | [max(): 最大値を取得](#max) |
| 3. | [int(): 整数型へキャスト](#int) || 23. | [min(): 最小値を取得](#min) |
| 4. | [bool(): bool 型へキャスト](#bool) || 24. | [all(): 全ての要素に対する判定](#all) |
| 5. | [float(): 64 bit 浮動小数点型へキャスト](#float) || 25. | [any(): いずれかの要素に対する判定](#any) |
| 6. | [str(): 文字列型へキャスト](#str) || 26. | [range(): 数のイミュータブルなシーケンスを指定長にて作成](#range) |
| 7. | [complex(): 複素数型へキャスト](#complex) || 27. | [sorted(): ソート結果を取得](#sorted) |
| 8. | [abs(): 絶対値の取得](#abs) || 28. | [slice(): スライスを取得](#slice) |
| 9. | [round(): 指定する桁を偶数丸め](#round) || 29. | [iter(): イテレータ作成](#iter) |
| 10. | [divmod(): 商と余りのタプルを取得](#divmod) || 30. | [reversed(): 要素を逆順に取り出すイテレータを取得](#reversed) |
| 11. | [sum(): 総和を取得](#sum) || 31. | [next(): イテレータの次の要素を取得](#next) |
| 12. | [pow(): べき乗計算](#pow) || 32. | [enumerate(): enumerate オブジェクトを取得](#enumerate) |
| 13. | [type(): オブジェクトの型を取得](#type) || 33. | [zip(): それぞれのイテラブルから要素を集めたイテレータを作成](#zip) |
| 14. | [isinstance(): 型の判定](#isinstance) || 34. | [map(): マップ型へキャスト](#map) |
| 15. | [dir(): オブジェクトの有効な属性のリストを返す](#dir) || 35. | [eval(): 文字列型の式の演算結果を取得](#eval) |
| 16. | [help(): ヘルプを開始／表示](#help) || 36. | [exec(): 文字列型の式を演算](#exec) |
| 17. | [list(): リスト型へキャスト](#list) || 37. | [open(): ファイルを開く](#open) |
| 18. | [tuple(): タプル型へキャスト](#tuple) || |
| 19. | [set(): 集合型へキャスト](#set) || |
| 20. | [dict(): 辞書型へキャスト](#dict) || |

---

## print(): 出力関数 <a class="anchor" id="print"></a>  
[目次に戻る](#0)  

`print(*objects, sep=" ", end="\n", file=sys.stdout, flush=False)`:  
`objects` を `sep` で区切りながらテキストストリーム `file`に表示し，最後に `end` を表示します．キーワード無しの引数は全て，`str()` がするように文字列に変換され，`sep` で区切られながらストリームに書き出され，最後に `end` が続きます．`sep` と `end` は両方とも文字列でなければなりません．

In [1]:
print("hello world")
print("hello", "world")
print("hello", "world", sep="-------")
print("hello", end="...")
print("world")

hello world
hello world
hello-------world
hello...world


---

## input(): 入力関数 <a class="anchor" id="input"></a>  
[目次に戻る](#0)  

`input(prompt)`:  
引数 `prompt` が存在すれば，それが末尾の開業を除いて標準出力に書き出されます．次に，この関数は入力から1行を読み込み，文字列に変換して（末尾の開業を除いて）返します．EOF が読み込まれたとき，`EOFError` が送出されます．

In [2]:
input()

 hello


'hello'

In [3]:
input("hello ")

hello  world


'world'

---

## int(): 整数型へキャスト <a class="anchor" id="int"></a>  
[目次に戻る](#0)  

`int(x)` または `int(x, base=10)`:  
数値または文字列 `x` から生成された整数オブジェクトを返します．引数が与えられない場合には 0 を返します．

In [4]:
int(1), int(-1), int("1"), int(0.999), int(float(1)), int()

(1, -1, 1, 0, 1, 0)

---

## bool(): bool 型へキャスト <a class="anchor" id="bool"></a>  
[目次に戻る](#0)  

`bool(x)`:  
ブール値を返します．`x` が偽または省略されている場合は `False` を返し，それ以外では `True` を返します．

In [5]:
(
    bool(1),
    bool(0),
    bool(-100),
    bool(100),
    bool(1+1j),
    bool(""),
    bool(" ")
)

(True, False, True, True, True, False, True)

`False` は以下の真理値判定手続きに基づくとのことで，これら以外は `True` (https://docs.python.org/ja/3.7/library/stdtypes.html#truth) ．

- 偽であると定義されている定数
- 数値型におけるゼロ
- 空のシーケンスまたはコレクション

---

## float(): 64 bit 浮動小数点型へキャスト <a class="anchor" id="float"></a>  
[目次に戻る](#0)  

`float(x)`:  
数または文字列 `x` から生成された浮動小数点数を返します．引数が文字列の場合，10進数を含んだ文字列にしてください．

In [6]:
float(1), float("-1"), float(1+2), float()

(1.0, -1.0, 3.0, 0.0)

---

## str(): 文字列型へキャスト <a class="anchor" id="str"></a>  
[目次に戻る](#0)  

`str(object="")` または `str(object=b"", encoding="utf-8", errors="strict")`:  
`object` の文字列版を返します．`object` が与えられなかった場合，空文字列が返されます．  
`encoding` か `errors` の少なくとも一方が与えられた場合，`object` は例えば bytes や bytearray の byte-like object でなければなりません．

In [7]:
str(123), str(), float(str(int(3.14)))

('123', '', 3.0)

---

## complex(): 複素数型へキャスト <a class="anchor" id="complex"></a>  
[目次に戻る](#0)  

`complex(real, imag)`:  
値 `real + imag*j` の複素数を返すか，文字列や数を複素数に変換します．第一引数が文字列なら，それが複素数と解釈され，この関数は第二引数なしで呼び出さなければなりません．第二引数は文字列であってはなりません．それぞれの引数は（複素数を含む）任意の数値型です．`imag` が省略された場合，標準の値はゼロで，このコンストラクタは `int` や `float` のような数値変換として働きます．両方の引数が省略された場合，`0j` を返します．引数に文字列を取る場合は，この文字列が複素数型ライクにする必要があり，かつ，実部と虚部の間の符号の前後に半角スペースを入れると `TypeError`になるので注意．

In [8]:
complex(1, 1), complex("1+1j")

((1+1j), (1+1j))

In [9]:
complex("1 + 1j")

ValueError: complex() arg is a malformed string

---

## abs(): 絶対値の取得 <a class="anchor" id="abs"></a>  
[目次に戻る](#0)  

`abs(x)`:  
`x` の絶対値を返します．引数が複素数である場合はノルム（複素平面上における複素ベクトルの長さ）を返します．

In [10]:
abs(5), abs(-5), abs(1+1j)

(5, 5, 1.4142135623730951)

最後の出力は $\sqrt{2} = 1.41421356...$ です．

---

## round(): 指定する桁を偶数丸め <a class="anchor" id="round"></a>  
[目次に戻る](#0)  

`round(number, ndigits)`:  
`number` の小数部を `ndigits` 桁に丸めた値を返します．`ndigits` が省略されたり `None` だった場合，入力値に最も近い整数を返します．  
`round()` をサポートする組み込み型では，値は 10  のマイナス `ndigits` 乗の倍数の中で最も近いものに丸められます．二つの倍数が同じだけ近いなら，偶数を選ぶ方に丸められます（偶数丸め，銀行丸め）．`ndigits` には任意の整数値が有効となります．返り値は `ndigits` が指定されていないか `None` の場合は整数，そうでなければ返り値は `number` と同じ型です．

In [11]:
round(1), round(3.14), round(0.999, 2), round(135, -1), round(0.5), round(0.501), round(0.501, 1)

(1, 3, 1.0, 140, 0, 1, 0.5)

---

## divmod(): 商と余りのタプルを取得 <a class="anchor" id="divmod"></a>  
[目次に戻る](#0)  

`divmod(a, b)`:  
複素数でない 2 つの数を引数として取り，整数の除去を行った時の商と余りからなるペアを返します．この返り値は，`(a // b, a % b)` と同じです．

In [12]:
divmod(5, 3)

(1, 2)

In [13]:
divmod(2**16 - 1, 17)

(3855, 0)

---

## sum(): 総和を取得 <a class="anchor" id="sum"></a>  
[目次に戻る](#0)  

`sum(iterable, start)`:  
`start` と `iterable` の要素を左から右へ合計し，総和を返します．`start` はデフォルトで `0` です．`iterable` の要素は通常は数値で，`start` の値は文字列であってはなりません．

In [14]:
sum((1,2)), sum([1, 2]), sum(range(5)), sum([0,0], 11)

(3, 3, 10, 11)

---

## pow(): べき乗計算 <a class="anchor" id="pow"></a>  
[目次に戻る](#0)  

`pow(x, y, z)`:  
`x` の `y` 乗を返し，これは `x**y` と等価です．`z` があれば，`x` の `y` 乗に対する `z` の余りを返し，これは `pow(x, y) % z` と同じ結果になりますが，前者の方がより効率よく計算されます．

In [15]:
pow(2, 8), pow(2, 3, 3), pow(2.718281828459045235360287471352, 3.141592653589793238462643j)

(256, 2, (-1+1.2246467991473532e-16j))

In [16]:
pow(2, 8) == 2**8, pow(2, 3, 3)==2**3 % 3

(True, True)

---

## type(): オブジェクトの型を取得 <a class="anchor" id="type"></a>  
[目次に戻る](#0)  

`type(object)` または `type(name, bases, dict)`:  
引数が 1 つだけの場合，`object` の型を返します．返り値は型オブジェクトで，一般に `object.__class__` によって返されるのと同じオブジェクトです．オブジェクトの型の判定には，`isinstance()` 関数を使うことが推奨されます．これはサブクラスを考慮するからです．  
引数が 3 つの場合，新しい型オブジェクトを返します．本質的には `class` 文の動的な形式です．`name` 文字列はクラス名で，`__name__` 属性になります．`bases` タプルは基底クラスの羅列で，`__bases__` 属性になります．`dict` はクラス本体の定義を含む名前空間で，標準の辞書にコピーされて `__dict__` 属性になります．

In [17]:
type(1), type(1.0), type(True), type(1j), type(None)

(int, float, bool, complex, NoneType)

In [18]:
new_type = type("x", (object,), {"a": 1})
new_type

__main__.x

In [19]:
new_type.a

1

In [20]:
type(new_type)

type

---

## isinstance(): 型の判定 <a class="anchor" id="isinstance"></a>  
[目次に戻る](#0)  

`isinstance(object, classinfo)`:  
`object` 引数が `classinfo` 引数のインスタンスであるか，サブクラスのインスタンスの場合に `True` を返します．`object` が与えられた方のオブジェクトでない場合，この関数は常に `False` を返します．`classinfo` が型オブジェクトのタプルの場合，`object` がそれらのいずれかのインスタンスであれば `True` を返します．`classinfo` が型や方からなるタプル，あるいは複数のタプルのいずれでもない場合，`TypeError` が送出されます．

In [21]:
a = "test"
b = 1
isinstance(a, str), isinstance(b, int), isinstance(b, float), isinstance(b, (str, int, float))

(True, True, False, True)

---

## dir(): オブジェクトの有効な属性のリストを返す <a class="anchor" id="dir"></a>  
[目次に戻る](#0)  

`dir(object)`:  
引数が無い場合，現在のローカルスコープにある名前のリストを返します．引数がある場合，そのオブジェクトの有効な属性のリストを返そうと試みます．

In [22]:
print(dir(int(-1)))

['__abs__', '__add__', '__and__', '__bool__', '__ceil__', '__class__', '__delattr__', '__dir__', '__divmod__', '__doc__', '__eq__', '__float__', '__floor__', '__floordiv__', '__format__', '__ge__', '__getattribute__', '__getnewargs__', '__gt__', '__hash__', '__index__', '__init__', '__init_subclass__', '__int__', '__invert__', '__le__', '__lshift__', '__lt__', '__mod__', '__mul__', '__ne__', '__neg__', '__new__', '__or__', '__pos__', '__pow__', '__radd__', '__rand__', '__rdivmod__', '__reduce__', '__reduce_ex__', '__repr__', '__rfloordiv__', '__rlshift__', '__rmod__', '__rmul__', '__ror__', '__round__', '__rpow__', '__rrshift__', '__rshift__', '__rsub__', '__rtruediv__', '__rxor__', '__setattr__', '__sizeof__', '__str__', '__sub__', '__subclasshook__', '__truediv__', '__trunc__', '__xor__', 'as_integer_ratio', 'bit_length', 'conjugate', 'denominator', 'from_bytes', 'imag', 'numerator', 'real', 'to_bytes']


In [23]:
int(1).__abs__()

1

In [24]:
print(dir())

['In', 'Out', '_', '_10', '_11', '_12', '_13', '_14', '_15', '_16', '_17', '_18', '_19', '_2', '_20', '_21', '_23', '_3', '_4', '_5', '_6', '_7', '_8', '__', '___', '__builtin__', '__builtins__', '__doc__', '__loader__', '__name__', '__package__', '__spec__', '_dh', '_i', '_i1', '_i10', '_i11', '_i12', '_i13', '_i14', '_i15', '_i16', '_i17', '_i18', '_i19', '_i2', '_i20', '_i21', '_i22', '_i23', '_i24', '_i3', '_i4', '_i5', '_i6', '_i7', '_i8', '_i9', '_ih', '_ii', '_iii', '_oh', 'a', 'b', 'exit', 'get_ipython', 'new_type', 'quit']


先ほど変数定義した，`a, b, new_type` が含まれていますね．

---

## help(): ヘルプを開始／表示 <a class="anchor" id="help"></a>  
[目次に戻る](#0)  

`help(object)`:  
`object` のヘルプを表示します．引数が与えられていない場合，対話的ヘルプシステムはインタプリタコンソール上で起動します．

In [25]:
help(abs)

Help on built-in function abs in module builtins:

abs(x, /)
    Return the absolute value of the argument.



In [26]:
a = None
help(a)

Help on NoneType object:

class NoneType(object)
 |  Methods defined here:
 |  
 |  __bool__(self, /)
 |      self != 0
 |  
 |  __repr__(self, /)
 |      Return repr(self).
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



---

## list(): リスト型へキャスト <a class="anchor" id="list"></a>  
[目次に戻る](#0)  

`list(iterable)`:  
`iterable` のリストを返します．

In [27]:
list([0, 1])

[0, 1]

In [28]:
list((0, 1))

[0, 1]

---

## tuple(): タプル型へキャスト <a class="anchor" id="tuple"></a>  
[目次に戻る](#0)  

`tuple(iterable)`:  
`iterable` のタプルを返します．

In [29]:
tuple([0, 1])

(0, 1)

---

## set(): 集合型へキャスト <a class="anchor" id="set"></a>  
[目次に戻る](#0)  

`set(iterable)`:  
オプションで `iterable` の要素を持つ，新しい `set` オブジェクトを返します．

In [30]:
set([0, 1, 1, 0, -1])

{-1, 0, 1}

In [31]:
set("""
はは　は　は　は　いい
""")

{'\n', '\u3000', 'い', 'は'}

In [32]:
set("I think that that that that that boy wrote is wrong.")

{' ',
 '.',
 'I',
 'a',
 'b',
 'e',
 'g',
 'h',
 'i',
 'k',
 'n',
 'o',
 'r',
 's',
 't',
 'w',
 'y'}

---

## dict(): 辞書型へキャスト <a class="anchor" id="dict"></a>  
[目次に戻る](#0)  

`dict(**kwarg)` または `dict(mapping, **kwarg)` または `dict(iterable, **kwarg)`: 新しい辞書を作成します．  
1 つ目は，`dict(key=value)` とすることにより辞書を作成します．  
2 つ目は，`dict(zip(key_list, value_list))` とすることにより辞書を作成します．  
3 つ目は，`dict([(key, value]))` とすることにより辞書を作成します．  

In [33]:
dict(map("1",1))

TypeError: 'int' object is not iterable

In [34]:
# dict(**kwarg)
dict(a=0, b=1, c=2)

{'a': 0, 'b': 1, 'c': 2}

In [35]:
# dict(mapping, **kwarg)
k = ["a", "b", "c"]
v = [0, 1, 2]
dict(zip(k, v))

{'a': 0, 'b': 1, 'c': 2}

In [36]:
# dict(iterable, **kwarg)
dict([("a", 0), ("b", 1), ("c", 2)])

{'a': 0, 'b': 1, 'c': 2}

---

## len(): 要素数／長さを取得 <a class="anchor" id="len"></a>  
[目次に戻る](#0)  

`len(s)`:  
オブジェクト `s` の長さ（要素の数）を返します．引数はシーケンスかコレクションです．

In [37]:
a = "test"
len(a)

4

In [38]:
a = [0, 1, 2, 3]
len(a)

4

In [39]:
a = {"a": 0, "b": 1}
len(a)

2

In [40]:
a = [[0, 1, 2, 3]]
len(a)

1

---

## max(): 最大値を取得 <a class="anchor" id="max"></a>  
[目次に戻る](#0)  

`max(iterable, key, default)` または `max(arg1, arg2, *args, key)`:  
`iterable` の中で最大の要素，または 2 つ以上の引数の中で最大のものを返します．

In [41]:
a = [0, 1, 2]
max(a)

2

In [42]:
max(0, 1, 2)

2

In [43]:
max(True, False, -1)

True

In [44]:
max("6", "1", 1)

TypeError: '>' not supported between instances of 'int' and 'str'

In [45]:
max(1, None)

TypeError: '>' not supported between instances of 'NoneType' and 'int'

大小関係が比較できない場合は，`TypeError` が送出されます．  
`None` はゼロでなく空なので，数値と比較することができません．

---

## min(): 最小値を取得 <a class="anchor" id="min"></a>  
[目次に戻る](#0)  

`min(iterable, key, default)` または `min(arg1, arg2, *args, key)`:  
`iterable` の中で最小の要素，または 2 つ以上の引数の中で最小のものを返します．性質や使い方は `max()` と同様です．

In [46]:
min(0, 1, 2, -1)

-1

---

## all(): 全ての要素に対する判定 <a class="anchor" id="all"></a>  
[目次に戻る](#0)  

`all(iterable)`:  
`iterable` の全ての要素が真，もしくは，空であれば `True` を返します．

In [47]:
(
    all([True]),
    all((0,1)),
    all([True is True, False is False, True is not False])
)

(True, False, True)

---

## any(): いずれかの要素に対する判定 <a class="anchor" id="any"></a>  
[目次に戻る](#0)  

`any(iterable)`:  
`iterable` のいずれかの要素が真ならば `True` を返します．`iterable` が空なら `False` を返します．

In [48]:
(
    any([True]),
    any((0,1)),
    any([True is True, False is False, True is not False])
)

(True, True, True)

---

## range(): 数のイミュータブルなシーケンスを指定長にて作成 <a class="anchor" id="range"></a>  
[目次に戻る](#0)  

`range(stop)` または `range(start, stop, step)`:  
range 型は，下図のイミュータブルなシーケンスを表し，一般に for ループにおいて特定の回数のループに使われます．引数は整数である必要があります．

In [49]:
range(0, 10)

range(0, 10)

In [50]:
list(range(0, 10))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [51]:
tuple(range(0, 5))

(0, 1, 2, 3, 4)

In [52]:
list(range(15, -3, -4))

[15, 11, 7, 3, -1]

---

## sorted(): ソート結果を取得 <a class="anchor" id="sorted"></a>  
[目次に戻る](#0)  

`sorted(iterable, *, key=None, reverse=False)`:  
`iterable` の要素を並び替えた新たなリストを返します．  
`key` には 1 引数関数を指定します．これは `iterable` の各要素から比較キーを展開するのにつかわれます．デフォルト値は `None` です．  
引数 `reverse` は真偽値です．`True` がセットされた場合，リストの要素は個々の比較が反転したものとして並び替えられます．

In [53]:
a = [1, 0, 3, -5, -10]
sorted(a)

[-10, -5, 0, 1, 3]

In [54]:
sorted(a, reverse=True)

[3, 1, 0, -5, -10]

---

## slice(): スライスを取得 <a class="anchor" id="slice"></a>  
[目次に戻る](#0)  

`slice(stop)` または `slice(start, stop, step)`:  
`range(start, stop, step)` で指定されるインデックスの集合を表す，スライスオブジェクトを返します．引数 `start` および `step` はデフォルトでは `None` です．スライスオブジェクトは読み出し専用の属性 `start, stop, step` を持ち，これらはたんに引数で使われた値を返します．

In [55]:
slice(10)

slice(None, 10, None)

In [56]:
slice(0, 10, 1)

slice(0, 10, 1)

In [57]:
a = [0, 1, 2, 3, 4]
a[slice(0, 3, 1)]

[0, 1, 2]

In [58]:
a = [0, 1, 2, 3, 4]
a[slice(0, 3, 1)]==a[0:3:1]

True

---

## iter(): イテレータ作成 <a class="anchor" id="iter"></a>  
[目次に戻る](#0)  

`iter(object, sentinel)`:  
イテレータオブジェクトを返します．第二引数があるかどうかで，第一引数の解釈は大きく異なります．第二引数が無い場合，`object` は反復プロトコル `__iter__()` メソッドか，シーケンスプロトコルをサポートする集合オブジェクトでなければなりません．  
第二引数 `sentinel` が与えられているなら，`object` は呼び出し可能オブジェクトでなければなりません．

In [59]:
iter([0,10])

In [60]:
iter((0,10))

In [61]:
list(iter([0,10]))

[0, 10]

In [62]:
list(iter(range(10)))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

---

## reversed(): 要素を逆順に取り出すイテレータを取得 <a class="anchor" id="reversed"></a>  
[目次に戻る](#0)  

`reversed(seq)`:  
要素を逆順に取り出すイテレータを返します．`seq` は `__reversed()` メソッドを持つか，シーケンス型プロトコルをサポートするオブジェクトでなければなりません．

In [63]:
reversed([0, 2, 1])

In [64]:
list(reversed([0, 2, 1]))

[1, 2, 0]

---

## next(): イテレータの次の要素を取得 <a class="anchor" id="next"></a>  
[目次に戻る](#0)  

`next(iterator, default)`:  
`iterator` の `__next__()` メソッドと呼び出すことにより，次の要素を取得します．イテレータが尽きている場合，`default` が与えられていればそれが返され，そうでなければ `StopIeration` が送出されます．

In [65]:
a = ["1", "2", "3"]
b = iter(a)
b

In [66]:
next(b)

'1'

In [67]:
next(b)

'2'

In [68]:
next(b)

'3'

In [69]:
next(b)

StopIteration: 

In [70]:
a = ["1", "2", "3"]
b = iter(a)
next(b, "end")

'1'

In [71]:
next(b, "end")

'2'

In [72]:
next(b, "end")

'3'

In [73]:
next(b, "end")

'end'

In [74]:
next(b, "end")

'end'

---

## enumerate(): enumerate オブジェクトを取得 <a class="anchor" id="enumerate"></a>  
[目次に戻る](#0)  

`enumerate(iterable, start=0)`: enumerate オブジェクトを返します．リスト化には，list 関数を用います．

In [75]:
a = ["a", "b", "c"]
b = enumerate(a)
b

例えばリスト化すると，インデックスと要素のペアをタプルで返します．

In [76]:
list(b)

[(0, 'a'), (1, 'b'), (2, 'c')]

---

## zip(): それぞれのイテラブルから要素を集めたイテレータを作成 <a class="anchor" id="zip"></a>  
[目次に戻る](#0)  

`zip(*iterables)`:  
それぞれのイテラブルから要素を集めたイテレータを作ります．  
この関数はタプルのイテレータを返し，その `i` 番目のタプルは引数シーケンスまたはイテラブルそれぞれの `i` 番目の要素を含みます．

In [77]:
a = [0, 1, 2]
b = ["b", "bb", "bbb"]
ab = zip(a, b)
ab

In [78]:
list(ab)

[(0, 'b'), (1, 'bb'), (2, 'bbb')]

---

## map(): マップ型へキャスト <a class="anchor" id="map"></a>  
[目次に戻る](#0)  

`map(function, iterable, ...)`:  
`function` を，結果を返しながら `iterable` の全ての要素に適用するイテレータを返します．追加の `iterable` 引数が渡された場合は，`function` はその数だけの引数を取らなければならず，全てのイテラブルから並行して撮られた要素に適用されます．

In [79]:
a = ["1", "2", "3"]
b = map(int, a)
b

In [80]:
list(b)

[1, 2, 3]

In [81]:
list(b)

[]

上記 2 個目の `list(b)` が `[]` となったのは，`b` がイテレータであり，1 回目の `list(b)` でインデックスが最後まで送られ，要素が全て取り出されたためです．予めリストとして違う文字変数に代入しておくことで，これは回避できます．

In [82]:
a = ["1", "2", "3"]
b = list(map(int, a))
b

[1, 2, 3]

In [83]:
b

[1, 2, 3]

In [84]:
b

[1, 2, 3]

---

## eval(): 文字列型の式の演算結果を取得 <a class="anchor" id="eval"></a>  
[目次に戻る](#0)  

`eval(expression, globals, locals)`:  
文字列とオプションの引数 `globals`, `locals` を取ります．オプション引数の `globals` は辞書で，`locals` はマッピングオブジェクトでなければなりません．

In [85]:
eval("1+2+3")

6

In [86]:
x = "hello"
y = "world"
eval("x + ' ' + y")

'hello world'

`eval` を用いると，このように文字列を式のように処理することができます．

---

## exec(): 文字列型の式を演算 <a class="anchor" id="exec"></a>  
[目次に戻る](#0)  

`exec(object, globals, locals)`:  
この関数は python コードの動的な実行をサポートします．`object` は文字列化コードオブジェクトでなければなりません．オプション引数の `globals` は辞書で，`locals` はマッピングオブジェクトでなければなりません．

In [87]:
exec("a=1")
a

1

In [88]:
x = "hello"
y = "world"
exec("z = x + ' ' + y")
z

'hello world'

---

## open(): ファイルを開く <a class="anchor" id="open"></a>  
[目次に戻る](#0)  

`open(file, mode="r", buffering=-1, encoding=None, errors=None, newline=None, closefd=True, opener=None)`:  
file を開き，対応するファイルオブジェクトを返します．ファイルを開くことができなければ， `OSError` が送出されます．  
`file` は path-like object （パスとして見なされるオブジェクト）で，開くファイルの (絶対または現在のワーキングディレクトリに対する相対) パス名を与えるものです．  
mode はオプションの文字列で，ファイルが開かれるモードを指定します．デフォルトは `'r'` で，読み込み用にテキストモードで開くという意味です．指定可能なモードは次の表の通りです．

| 文字 | 意味 |
| ---- | ---- |
| 'r'  | 読み込み用に開く (デフォルト) |
| 'w'  | 書き込み用に開き、まずファイルを切り詰める |
| 'x'  | 排他的な生成に開き、ファイルが存在する場合は失敗する |
| 'a'  | 書き込み用に開き、ファイルが存在する場合は末尾に追記する |
| 'b'  | バイナリモード |
| 't'  | テキストモード (デフォルト) |
| '+'  | ディスクファイルを更新用に開く (読み込み／書き込み) |

公式ドキュメント (https://docs.python.org/ja/3.7/library/functions.html#open) には，網羅的な説明がなされていますが，例を見た方が良いと思うので，例を以下に示します．  

まずは，ファイルを書き込み用に新規で開き，適当な値を代入していきます．

In [89]:
# file path
file = "test.txt"

# get file object @ mode="w"
f = open(file, mode="w")
f

<_io.TextIOWrapper name='test.txt' mode='w' encoding='cp932'>

In [90]:
type(f)

_io.TextIOWrapper

これで test.txt が書き込み専用で新規に作成され，`f` をこれのファイルオブジェクトとして定義しました．変数の名称は何でも構いません．  
次に `write()` メソッドを用いて書き込んでみます．`write()` の引数は文字列のみ有効なので，もし数値を入れたい場合は，f-string を用いるか，`str()` にて文字列型にキャストしましょう．

In [91]:
f.write("aaaaa")
f.write(str(3.14))
f.write("あああああ")
f.write(f"{list(range(5))}")

15

`open()` したファイルオブジェクト `f` は，`close()` メソッドを用いて閉じる必要があります．閉じることで，ファイルへの書き込みが完了します．

In [92]:
f.close()

出力結果を見るために，今度は読み込んでみましょう．`mode=="r"` として読み込み専用で開き，`read()` メソッドにて文字列を取得しましょう．

In [93]:
# open
file = "test.txt"
f = open(file, mode="r")

# read text
txt = f.read()

# file close
f.close()

# show
txt

'aaaaa3.14あああああ[0, 1, 2, 3, 4]'

上記のように，`write()` メソッドにて指定した文字列が改行なしでテキストファイルに入力されました．  
ファイルオブジェクトの open / close について，上記の書き方だと close を忘れる危険性があるので，`with` ブロックというのを用いる方が一般的です．`with open() as f` のように記載します．これを用いることで，`with` ブロックを抜ける際に，`open()` したファイルオブジェクトが自動的に `close()` されます．

In [94]:
file = "test.txt"
with open(file, mode="w") as f:
    f.write("with block test")

with open(file, mode="r") as f:
    txt2 = f.read()

txt2

'with block test'